src/main.rs

```diff
@@ -0,0 +1,116 @@
+const XAMZSECURITYTOKEN: &str = "x-amz-security-token";
+const XAMZDATE: &str = "x-amz-date";
+
+use simple_proxy::proxy::error::MiddlewareError;
+use simple_proxy::proxy::middleware::MiddlewareResult::Next;
+use simple_proxy::proxy::middleware::{Middleware, MiddlewareResult};
+use simple_proxy::proxy::service::{ServiceContext, State};
+
+pub struct AwsSignatureHeaders {
+    region: String
+}
+
+impl AwsSignatureHeaders {
+    pub fn new(region: String) -> Self {
+        AwsSignatureHeaders{
+            region: region
+        }
+    }
+}
+
+impl Middleware for AwsSignatureHeaders {
+    fn name() -> String {
+        String::from("AwsSignatureHeaders")
+    }
+
+    fn before_request(
+        &mut self,
+        req: &mut Request<Body>,
+        _context: &ServiceContext,
+        _state: &State,
+    ) -> Result<MiddlewareResult, MiddlewareError> {
+        let (amzdate, datestamp) = aws_signature_builder::get_utc_aws_signature_datestrings();
+        let mut headers = ::std::collections::HashMap::new();
+        for (key, value) in req.headers().iter() {
+            headers.insert(String::from(key.as_str()), String::from(value.to_str().unwrap()));
+        }
+        let port = match req.uri().port_part() {
+            Some(x) => Some(x.as_u16()),
+            None => None,
+        };
+        let host: String = req.uri().host().unwrap().to_string();
+        let host_parts: Vec<&str> = host.split(".").collect();
+        let provider = DefaultCredentialsProvider::new().unwrap();
+        let credentials = provider.credentials().wait().unwrap();
+        let new_headers = aws_signature_builder::generate_aws_signature_headers(
+            req.uri().query().unwrap().to_string(),
+            headers,
+            port,
+            req.uri().host().unwrap().to_string(),
+            amzdate,
+            req.method().to_string(),
+            Vec::new(),
+            None,
+            false,
+            datestamp,
+            host_parts[0].to_string(),
+            self.region.clone(),
+            credentials.aws_access_key_id().to_string(),
+            credentials.aws_secret_access_key().to_string(),
+            req.uri().path().to_string());
+        if new_headers.contains_key(XAMZCONTENTSHA256) {
+            req.headers_mut().insert(XAMZCONTENTSHA256,
+                HeaderValue::from_str(&new_headers[XAMZCONTENTSHA256]).unwrap());
+        }
+        if new_headers.contains_key(XAMZSECURITYTOKEN) {
+            req.headers_mut().insert(XAMZSECURITYTOKEN,
+                HeaderValue::from_str(&new_headers[XAMZSECURITYTOKEN]).unwrap());
+        }
+        if new_headers.contains_key(XAMZDATE) {
+            req.headers_mut().insert(XAMZDATE,
+                HeaderValue::from_str(&new_headers[XAMZDATE]).unwrap());
+        }
+        if new_headers.contains_key("Authorization") {
+            req.headers_mut().insert(AUTHORIZATION,
+                HeaderValue::from_str(&new_headers["Authorization"]).unwrap());
+        }
```

Comment:
It might be nice to have a reference to the docs or a comment to understand when someone should/shouldn't expect these headers or exist? Unless you think that's unnecessary/unimportant.

Reply:
I think this is related to [this earlier thread about the headers](https://github.com/sverch/aws-signature-proxy/pull/1#discussion_r337236271).  Much of this code was copy/pasted from the AWS docs into awscurl, and then translated line by line into rust, so that understanding might not even be possible.  I'll say all that in a comment here as well.